kedro/io/data_catalog.py

```diff
@@ -271,22 +313,40 @@ class to be loaded is specified with the key ``type`` and their
 
         layers: dict[str, set[str]] = defaultdict(set)
         for ds_name, ds_config in catalog.items():
-            ds_layer = ds_config.pop("layer", None)
-            if ds_layer is not None:
-                layers[ds_layer].add(ds_name)
-
-            ds_config = _resolve_credentials(ds_config, credentials)
-            data_sets[ds_name] = AbstractDataSet.from_config(
-                ds_name, ds_config, load_versions.get(ds_name), save_version
-            )
+            # Assume that any name with "}" in it is a dataset factory to be matched.
```

Comment:
I would've rewritten this for readability as follows:

1. Create a factory method for creating a dataset out of `def create_dataset(cls, ds_name, ds_config):`
2. Convert this single for loop with an if in it into two Python comprehensions for the two different cases